### PR TITLE
Kafka Aggregate message dispatching

### PR DESCRIPTION
**Pull request by dmdashenkov**
In this PR we introduce `KAggregateRepository` - an `abstract` class extended from `AggregateRepository`, which changes the way the `Command`s, `Event`s and `Rejections` are dispatched by the repository.
There is also the `KAggregatePartRepository` for `AggregatePart` message dispatching.

`KAggregateRepository` uses Endpoint delivery implementations that publish the messages into a dedicated Kafka topic. Then, a Kafka Streams topology processes the messages grouping them by the entity ID, so that the messages to one `Aggregate` instance are processed within a single machine/thread.

Example:
```java
// --- TaskRepository.java ---
public class TaskRepository extends KAggregateRepository<TaskId, TaskAggregate> { 

    public TaskRepository(Properties streamConfig, KafkaWrapper kafka) {
        super(streamConfig, kafka);
    }
}
```
Then register the `TaskRepository` in your `BoundedContext` instance. No additional configuration is required.
Note that the `streamConfig` argument in the `KAggregateRepository` constructor is the Kafka Streams configuration. It may look as follows:
```properties
# 2 bootstrap servers, comma separated. Specify as many as possible.
bootstrap.servers=my.kafka.broker.host1.io,my.kafka.broker.host2.io

# Some other configuration parameters (optionally).
```
The property `bootstrap.servers` is required for the Kafka Streams topology to start successfully.

Note that the dispatching process requires that a specific Kafka topic is created before the application is started. See `KAggregateRepository` for the details.

### CODE REVIEW

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

**Comment by armiol** [comment]
@dmdashenkov please have an example in the PR description. Or something. It's going to be very helpful to those, who look for a code sample. They ain't gonna look through 100+ of changed files for that.

File: api-java/src/main/java/io/spine/examples/todolist/context/BoundedContextFactory.java [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+     *
+     * <p>Creates a new {@link InMemoryStorageFactory}.
+     */
+    @SuppressWarnings("Guava") // Spine Java 7 API.
+    private static final Supplier<StorageFactory> IN_MEM_STORAGE_SUPPLIER =
+            () -> InMemoryStorageFactory.newInstance(BOUNDED_CONTEXT_NAME, false);
+
+    private final StorageFactorySwitch storageFactorySwitch;
+
+    public static BoundedContextName getDefaultName() {
```
Comment (armiol) [inline api-java/src/main/java/io/spine/examples/todolist/context/BoundedContextFactory.java]: Please document, as long as you modify. Also, are there any other names?

File: api-java/src/main/java/io/spine/examples/todolist/context/BoundedContextFactory.java [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+    /**
+     * A trigger method for extending {@link #create()} method logic.
+     *
+     * <p>This method is called after an instance on {@link BoundedContext} is created and all
+     * the repositories are registered.
+     *
+     * @param bc the created {@link BoundedContext}
+     * @implSpec Performs no action by default.
+     */
+    protected void onCreateBoundedContext(BoundedContext bc) {
```
Comment (armiol) [inline api-java/src/main/java/io/spine/examples/todolist/context/BoundedContextFactory.java]: This is weird. `on...Created` would make more sense.

Also, it all looks weird in a context of a particular `Todo List` application.

File: api-java/src/main/java/io/spine/examples/todolist/context/BoundedContextFactory.java [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+        // NoOp
+    }
+
+    private static EventBus.Builder createEventBus(
+            StorageFactory storageFactory,
+            AggregateRepository<LabelId, LabelAggregate> labelRepo,
```
Comment (armiol) [inline api-java/src/main/java/io/spine/examples/todolist/context/BoundedContextFactory.java]: This is a bad design. With three more repos we're going to redo everything.

File: deployment/local-kafka/src/main/java/io/spine/examples/todolist/server/KafkaBoundedContextFactory.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+public final class KafkaBoundedContextFactory extends BoundedContextFactory {
+
+    private static final String KAFKA_STREAMS_PROPS_PATH = "config/kafka-streams.properties";
+    private static final Properties streamsConfig;
+    private static final StorageFactorySwitch storageFactorySwitch;
+    private static final KafkaWrapper kafka;
+
+    static {
+        final Properties producerConfig = loadConfig("config/kafka-producer.properties");
+        final Properties consumerConfig = loadConfig("config/kafka-consumer.properties");
```
Comment (armiol) [inline deployment/local-kafka/src/main/java/io/spine/examples/todolist/server/KafkaBoundedContextFactory.java]: I don't get how one path is a constant (see line:74) and these two aren't. Please extract.

File: deployment/local-kafka/src/main/java/io/spine/examples/todolist/server/KafkaBoundedContextFactory.java [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+        final Duration storagePollAwait = Duration.of(50, MILLIS);
```
Comment (armiol) [inline deployment/local-kafka/src/main/java/io/spine/examples/todolist/server/KafkaBoundedContextFactory.java]: I think the settings that you use (such as poll await time and strong consistency mode) require some words somewhere, maybe, in a class description. E.g. what were the intention, how it works etc.

File: deployment/local-kafka/src/main/java/io/spine/examples/todolist/server/repository/TaskKRepository.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+import io.spine.examples.todolist.TaskId;
+import io.spine.examples.todolist.c.aggregate.TaskAggregateRoot;
+import io.spine.examples.todolist.c.aggregate.TaskPart;
+import io.spine.server.aggregate.KAggregatePartRepository;
+import io.spine.server.storage.kafka.KafkaWrapper;
+
+import java.util.Properties;
+
+/**
+ * An implementation of a repository for {@link TaskPart} derived from
```
Comment (armiol) [inline deployment/local-kafka/src/main/java/io/spine/examples/todolist/server/repository/TaskKRepository.java]: Please do not duplicate the declaration of a class. Say something about its features or responsibility. 

And it's not only here.

File: kafka-storage/src/main/java/io/spine/server/aggregate/KAggregatePartRepository.java [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+import java.util.Collections;
+import java.util.Properties;
+import java.util.Set;
+import java.util.function.Consumer;
+
+/**
+ * An {@link AggregatePartRepository} which applies the dispatched events through a Kafka Streams
+ * topology.
+ *
+ * <p>This class is an analogy of {@link KAggregateRepository} for {@link AggregatePart}.
```
Comment (armiol) [inline kafka-storage/src/main/java/io/spine/server/aggregate/KAggregatePartRepository.java]: This is not an "analogy".

`This class plays the same role to ... as ... to ...`.

File: kafka-storage/src/main/java/io/spine/server/aggregate/KAggregatePartRepository.java [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+import io.spine.server.entity.EntityClass;
+import io.spine.server.storage.kafka.KafkaWrapper;
+
+import java.util.Collections;
+import java.util.Properties;
+import java.util.Set;
+import java.util.function.Consumer;
+
+/**
+ * An {@link AggregatePartRepository} which applies the dispatched events through a Kafka Streams
```
Comment (armiol) [inline kafka-storage/src/main/java/io/spine/server/aggregate/KAggregatePartRepository.java]: `through` requires serious explanation.

File: kafka-storage/src/main/java/io/spine/server/aggregate/KAggregatePartRepository.java [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+    protected KAggregatePartRepository(Properties streamConfig, KafkaWrapper kafka) {
+        super();
+        this.kafka = kafka;
+        KafkaAggregateLoading.start(this, streamConfig);
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @return always empty set
```
Comment (armiol) [inline kafka-storage/src/main/java/io/spine/server/aggregate/KAggregatePartRepository.java]: That's illegal. If that is dispatched to someone, it should return their ID(s).

File: kafka-storage/src/main/java/io/spine/server/aggregate/KAggregatePartRepository.java [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+     * @implSpec Publishes the given {@linkplain RejectionEnvelope rejection} into
+     * the {@code spine.server.aggregate.messages} Kafka topic as a key-value pair:
+     * <pre>
+     * {@linkplain #key() Repository type} -> {@link RejectionEnvelope#getOuterObject()}.
+     * </pre>
+     */
+    @Override
+    public Set<I> dispatchRejection(RejectionEnvelope envelope) {
+        KafkaAggregateLoading.dispatchMessage(this, envelope);
```
Comment (armiol) [inline kafka-storage/src/main/java/io/spine/server/aggregate/KAggregatePartRepository.java]: `KafkaAggregateLoading` would probably look better with `KafkaAggregateLoader` name.

File: kafka-storage/src/main/java/io/spine/server/aggregate/KAggregatePartRepository.java [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+     * @param task the {@link Consumer} to execute
+     * @param argument the {@code task} argument
+     * @param <E> the type of the {@code argument} for the {@code task}
+     */
+    @SuppressWarnings("DuplicateStringLiteralInspection")
+        // OK for the log messages. Duplicate in `KAggregateRepository`.
+    private <E extends MessageEnvelope<?, ? ,?>> void logErrors(Consumer<E> task, E argument) {
+        try {
+            task.accept(argument);
+            log().info("Acknowledged {} (ID: {}).",
```
Comment (armiol) [inline kafka-storage/src/main/java/io/spine/server/aggregate/KAggregatePartRepository.java]: It doesn't look like `info`. There is going to be lots and lots of those. Please review the cases like that.

File: kafka-storage/src/main/java/io/spine/server/aggregate/KAggregateRepository.java [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+     * @implSpec Publishes the given {@linkplain EventEnvelope event} into
+     * the {@code spine.server.aggregate.messages} Kafka topic as a key-value pair:
+     * <pre>
+     * {@linkplain #key() Repository type} -> {@link EventEnvelope#getOuterObject()}.
+     * </pre>
+     */
+    @Override
+    public final Set<I> dispatchEvent(EventEnvelope envelope) {
+        KafkaAggregateLoading.dispatchMessage(this, envelope);
+        return Collections.emptySet();
```
Comment (armiol) [inline kafka-storage/src/main/java/io/spine/server/aggregate/KAggregateRepository.java]: The same comments apply, as previously.

File: kafka-storage/src/main/java/io/spine/server/aggregate/KAggregateRepository.java [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+     * {@linkplain MessageEnvelope message argument} and catches and logs all the runtime
+     * exceptions thrown by the {@code task}.
+     *
+     * @param task the {@link Consumer} to execute
+     * @param argument the {@code task} argument
+     * @param <E> the type of the {@code argument} for the {@code task}
+     */
+    @SuppressWarnings("DuplicateStringLiteralInspection")
+        // OK for the log messages. Duplicate in `KAggregateRepository`.
+    private <E extends MessageEnvelope<?, ? ,?>> void logErrors(Consumer<E> task, E argument) {
```
Comment (armiol) [inline kafka-storage/src/main/java/io/spine/server/aggregate/KAggregateRepository.java]: Please think of another name. This one does not correspond to the description and/or implementation.

File: kafka-storage/src/main/java/io/spine/server/aggregate/KafkaAggregateLoading.java [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+     * {@code KAggregateRepository}.
+     */
+    private static final Topic AGGREGATE_MESSAGES = ofValue("spine.server.aggregate.messages");
+
+    private KafkaAggregateLoading() {
+        // Prevent utility class instantiation.
+    }
+
+    /**
+     * Starts the Aggregate loading for the given {@link KafkaAggregateRepository}.
```
Comment (armiol) [inline kafka-storage/src/main/java/io/spine/server/aggregate/KafkaAggregateLoading.java]: No idea what it means.

File: kafka-storage/src/main/java/io/spine/server/aggregate/KafkaAggregateRepository.java [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+ * the Aggregates.
+ *
+ * <p>Note that all the methods of this interface are {@linkplain Internal internal}. Please use
+ * the basic {@link AggregateRepository} API instead.
+ *
+ * @author Dmytro Dashenkov
+ * @see KAggregateRepository
+ * @see KAggregatePartRepository
+ */
+interface KafkaAggregateRepository {
```
Comment (armiol) [inline kafka-storage/src/main/java/io/spine/server/aggregate/KafkaAggregateRepository.java]: This needs to be explained. Why it exists. Why the methods are named `...Now` with the description, saying nothing about this "Now".

File: kafka-storage/src/main/java/io/spine/server/kafka/KafkaStreamsConfigs.java [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+    private KafkaStreamsConfigs() {
+        // Prevent utility class instantiation.
+    }
+
+    /**
+     * Appends
+     * the {@link org.apache.kafka.streams.StreamsConfig#APPLICATION_ID_CONFIG application.id}
+     * configuration to the given properties.
+     *
+     * <p>Note: this method creates a defencive copy of the passed {@link Properties} instead of
```
Comment (armiol) [inline kafka-storage/src/main/java/io/spine/server/kafka/KafkaStreamsConfigs.java]: `defencive` is a typo.

File: kafka-storage/src/main/java/io/spine/server/aggregate/KafkaAggregateRepository.java (hunk 15 above)
Comment (dmdashenkov) [reply kafka-storage/src/main/java/io/spine/server/aggregate/KafkaAggregateRepository.java]: Done

File: kafka-storage/src/main/java/io/spine/server/aggregate/KAggregatePartRepository.java (hunk 8 above)
Comment (dmdashenkov) [reply kafka-storage/src/main/java/io/spine/server/aggregate/KAggregatePartRepository.java]: Added a link to the proper explanation.

File: kafka-storage/src/main/java/io/spine/server/aggregate/KafkaAggregateMessageDispatcher.java [hunk 17]
```diff
@@ -0,0 +1,235 @@
...
+                                                                       id, message);
+        kafkaProducer.send(record);
+    }
+
+    private void buildTopology(KStream<I, Message> stream) {
+        stream.foreach(this::doDispatch);
+    }
+
+    @SuppressWarnings({"IfStatementWithTooManyBranches", "ChainOfInstanceofChecks"})
+        // OK for this method as we want a *single* processing point for any kind of message.
```
Comment (armiol) [inline kafka-storage/src/main/java/io/spine/server/aggregate/KafkaAggregateMessageDispatcher.java]: Why couldn't we define a generic way here and deal with the exact types in the implementation of this class?

File: kafka-storage/src/main/java/io/spine/server/storage/kafka/Topic.java [hunk 18]
```diff
@@ -61,63 +63,72 @@ static Topic forRecordOfType(Class<?> ofType) {
...
     static Topic ofValue(String topicValue) {
         return new ValueTopic(topicValue);
     }
 
     /**
-     * Creates a topic for all the events in the system.
+     * Creates an instance of {@code Topic} for the aggregate messages of the given aggregate state
+     * type.
+     *
+     * <p>This topic has the name equal to the {@link TypeUrl#getTypeName()} of the given type url.
```
Comment (armiol) [inline kafka-storage/src/main/java/io/spine/server/storage/kafka/Topic.java]: `url` is capitalized.

File: kafka/libs/kafka_2.12-1.0.0-javadoc.jar.asc [hunk 19]
```diff
@@ -0,0 +1,17 @@
+-----BEGIN PGP SIGNATURE-----
```
Comment (armiol) [inline kafka/libs/kafka_2.12-1.0.0-javadoc.jar.asc]: Not sure why the binaries are here.

File: api-java/src/main/java/io/spine/examples/todolist/context/BoundedContextFactory.java [hunk 20]
```diff
@@ -0,0 +1,266 @@
...
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.examples.todolist.context;
```
[the remark this answers is not in the document]
Comment (dmdashenkov) [reply api-java/src/main/java/io/spine/examples/todolist/context/BoundedContextFactory.java]: Created issue #39.

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

**Pull request merged**